public/app/features/trails/Integrations/interpolation.ts

```diff
@@ -0,0 +1,56 @@
+import { AdHocVariableFilter, DataSourceInstanceSettings } from '@grafana/data';
+import { PromQuery, PrometheusDatasource } from '@grafana/prometheus';
+import { getDataSourceSrv } from '@grafana/runtime';
+import { AdHocFiltersVariable, SceneVariable } from '@grafana/scenes';
+import { DataQuery, DataSourceRef } from '@grafana/schema';
+import { DashboardModel } from 'app/features/dashboard/state';
+import { DashboardScene } from 'app/features/dashboard-scene/scene/DashboardScene';
+
+export function interpolateVariables(
+  dashboard: DashboardModel | DashboardScene,
+  dsInstanceSettings: DataSourceInstanceSettings,
+  queries: DataQuery[]
+) {
+  const promDataSourceForUtility = new PrometheusDatasource({ ...dsInstanceSettings, jsonData: {} });
+
+  return queries.filter(isPromQuery).map((query) => {
+    const queryDataSourceUid = getInterpolatedDataSourceUid(query?.datasource);
+    return promDataSourceForUtility.applyTemplateVariables(query, {}, getAdHocFilters(dashboard, queryDataSourceUid));
+  });
+}
+
+function getAdHocFilters(dashboard: DashboardModel | DashboardScene, queryDataSourceUid: string | undefined) {
+  const adhocFilters: AdHocVariableFilter[] = [];
+  if (dashboard instanceof DashboardModel) {
+    dashboard.getVariables().forEach((variable) => {
+      if (variable.type === 'adhoc') {
+        const variableDataSourceUid = getInterpolatedDataSourceUid(variable.datasource);
+        if (variableDataSourceUid === queryDataSourceUid) {
+          variable.filters.forEach((filter) => adhocFilters.push(filter));
+        }
+      }
+    });
+  } else {
+    dashboard.state.$variables?.state.variables.filter(isAdHocFiltersVariable).forEach((variable) => {
+      const variableDataSourceUid = getInterpolatedDataSourceUid(variable.state.datasource);
+      if (variableDataSourceUid === queryDataSourceUid) {
+        variable.state.filters.forEach((filter) => adhocFilters.push(filter));
+      }
+    });
+  }
+
+  return adhocFilters;
+}
```

Comment:
This wont work in scenes, 

Implemented a simple scenes version here: https://github.com/grafana/grafana/pull/86779

I think we can skip support for old dashboards, it's going away soon anyway and this use case from dashboards is a bit experimental anyway

Reply:
It would certainly simplify things to skip support for the old dashboards.
Are we expecting the change to scenes dashboards to make it into 11.0.x?

Reply:
Dropped support for legacy dashboards and took advantage of the async and scenes objects.
Everything feels *much* cleaner now.